File: 215.数组中的第K个最大元素/quickSort.java
import java.util.Random;

class Solution {
    Random random = new Random();

    public int findKthLargest(int[] nums, int k) {      //使用快排
        return quickSelect(nums,0,nums.length-1,nums.length-k);
    }
    
    public int quickSelect(int[] a,int l, int r, int index){        //in：数组，左边界，右边界，目标位置；
        int q = randomPartition(a,l,r);
        if ( q == index){
            return a[q];
        }
        else{
            return q<index?quickSelect(a,q+1,r,index):quickSelect(a,l,q-1,index);
        }
    }
    
    public int randomPartition(int[] a, int l, int r){      //从数组中随机选出一个值，并为接下来的方便移到最右边，然后给到partition函数进行位置交换确定该数的下标；in：数组， 左右值
        int i = random.nextInt(r-l+1)+l;
        swap(a,i,r);
        return partition(a,l,r);
    }
    
    public int partition(int[] a,int l , int r){        //对数组l-r范围进行交换，大的移到右边
        int x=a[r],i=l-1;
        for(int j=l;j<r;j++){
            if(a[j]<x){
                swap(a,++i,j);
            }
        }
        swap(a,i+1,r);
        return(i+1);
    }
    
    public void swap(int[] a,int i,int j){      //交换函数
        int temp=a[i];
        a[i]=a[j];
        a[j]=temp;
    }
}



/*
快速排序，典型的分治算法
使用方法：分解，将数组划分为两个子数组，a[l...q-1]比a[q]小，a[q+1...r]比a[q]大；解决，通过递归，对两个子数组进行快速排序
每次划分后，能确定一个元素的正确位置，如果小于index则在右边继续找，反之在左边，若正好等于index则为需要的值



执行用时：2 ms, 在所有 Java 提交中击败了91.38%的用户
内存消耗：38.7 MB, 在所有 Java 提交中击败了59.98%的用户

时间复杂度：O(n)，期望为线性的选择算法
空间复杂度：：O(logn)，递归栈使用空间为logn
*/

File: 912.排序数组/quickSort.py
class Solution:
    def swap(self,a,i,j):
        temp=a[i]
        a[i]=a[j]
        a[j]=temp
    def quickSort(self,a,l,r):
        if l>=r:
            return
        mid = self.partition(a,l,r)
        self.quickSort(a,l,mid-1)
        self.quickSort(a,mid+1,r)

    def partition(self,a,l,r):
        x=random.randint(l,r)
        self.swap(a,x,r)
        i=l-1
        for j in range(l,r):
            if a[j]>=a[r]:
                continue
            i+=1
            self.swap(a,i,j)
        i+=1
        self.swap(a,i,r)
        return i

    def sortArray(self, nums: List[int]) -> List[int]:
        self.quickSort(nums,0,len(nums)-1)
        return nums
    
'''
快速排序
如果数大，则i不变，i一定要找到一个小的数，才变，这样让i的那一边都是比目标数小




执行用时：708 ms, 在所有 Python3 提交中击败了9.91%的用户
内存消耗：19.9 MB, 在所有 Python3 提交中击败了42.72%的用户
时间：O(nlogn)，基于随机选择目标的快速排序
空间：O(h)，h为快速排序递归所用的树的层数，最坏是n，即每一个数占用一层；最好是O(logn)，为平衡树
'''